File: 设计模式/程序/10-DecoratorPattern/DecoratorPattern(python)/ShapeDecorator.py
from Shape import *

# 形状的装饰器基类
class ShapeDecorator(Shape):
    @abc.abstractmethod
    def __init__(self, shape):
        self.shape = shape
    

# 装饰器实现类
class ColorShape(ShapeDecorator):
    def __init__(self, shape, color):
        ShapeDecorator.__init__(self, shape)
        self.color = color
    # 实现接口
    def draw(self):
        # 除了原来的绘画功能，还增加了颜色
        self.shape.draw()
        print("fill with", self.color, "color")

# 还可以继续拓展别的形状装饰器...

File: 设计模式/程序/10-DecoratorPattern/DecoratorPattern(C#)/ShapeDecorator.cs
using System;

namespace DecoratorPattern
{
    public abstract class ShapeDecorator : Shape
    {   
        protected Shape shape;
        public ShapeDecorator(Shape shape)
        {
            this.shape = shape;
        }
        public virtual void draw()
        {
            shape.draw();
        }
    } 
    
    public class ColorShape: ShapeDecorator
    {
        protected string color;
        public ColorShape(Shape shape, string color): base(shape)
        {
            this.color = color;
        }
        public override void draw()
        {
            shape.draw();
            Console.WriteLine("With {0} Color", this.color);
        }
    }
}